### PR TITLE
Ultima develop code merge

### PR DESCRIPTION
**Pull request by dror27**
This pull request includes Ultima Genomics code contributed to the public IGV version.

Code mainly has to do with additional rendering features for Flow Based tracks, specifically graphic indications of INDEL qualities.

NOTE: please hold on reviewing this pull request, pending out-of-bound communication,.

### CODE REVIEW

**Comment by ilyasoifer** [comment]
@jrobinso - this is the PR related to #1393. Can you suggest how to move forward with that? I am also attaching the documentation of the changes: 
[IGV Supplemental.pdf](https://github.com/igvteam/igv/files/13197381/IGV.Supplemental.pdf)

**Comment by jrobinso** [comment]
@ilyasoifer Apologies for the delay,  I am on vacation returning Nov 19.  Will respond further via email.

**Comment by jrobinso** [comment]
WRT to next steps,  we will need a  test case we can explore ourselves.   Also, eventually we would request the documentation in Markdown, as that is the language used for our documentation system (https://igv.org/doc/desktop/).  Thanks.

**Comment by ilyasoifer** [comment]
@jrobinso - thanks for the feedback. We will add the test data (I assume to `test/data`). Could you please point me to the repository that contains the documentation so I can see how the md documents you are using are structured?

**Comment by dror27** [comment]
> @jrobinso - thanks for the feedback. We will add the test data (I assume to `test/data`). Could you please point me to the repository that contains the documentation so I can see how the md documents you are using are structured? Enjoy your vacation!

flow_renderer.sam added to test/data/sam. Once @ilyasoifer adds doc we can point to specific reads & locations for each new feature. Data is around chr1:10000

**Comment by helgathorv** [comment]
@ilyasoifer Our new doc repo is not yet public, but here are some sample markdown files from the _User Guide > Tracks and Data Types > Alignments_. Let me know if you have any questions.
[SampleIGVDoc.zip](https://github.com/igvteam/igv/files/13350783/SampleIGVDoc.zip)

**Comment by helgathorv** [comment]
@ilyasoifer The IGV doc repo is now public at https://github.com/igvteam/igv-docs-website

**Comment by ilyasoifer** [comment]
Thanks @helgathorv for these examples. Should we create a PR for the documentation in that repository or attach it to this PR?

**Comment by jrobinso** [comment]
@ilyasoifer As this is just a single page of doc attaching it here is fine.

**Comment by ilyasoifer** [comment]
@jrobinso and @helgathorv I added to this PR a folder called `doc` that contains a markdown file `igv.md`. Since the document also contains images, I thought it is more convenient to add it this way. Let me know if you would like me to otherwise re-arrange.

**Comment by ilyasoifer** [comment]
@dror27 - I think our code contains a small bug similar to what we encountered in broadinstitute/gatk:#8337. 
Specifically, the code automatically recognizes data that contains `tp` tag as Ultima. However, minimap2 aligner apparently also uses `tp` tag for a different reason. 
Could you implement something similar to what I did in that PR? Thanks!

**Comment by dror27** [comment]
> @dror27 - I think our code contains a small bug similar to what we encountered in [broadinstitute/gatk#8337](https://github.com/broadinstitute/gatk/pull/8337). Specifically, the code automatically recognizes data that contains `tp` tag as Ultima. However, minimap2 aligner apparently also uses `tp` tag for a different reason. Could you implement something similar to what I did in that PR? Thanks!

Logic added to screen flow alignments based on read-group presence and fields

**Comment by jrobinso** [comment]
Hi all, apologies for the lack of response, I've been delayed by travel and some pressing issues but will get to it ASAP.

**Comment by ilyasoifer** [comment]
No worries, have a great holiday!

**Comment by jrobinso** [comment]
Hi all,  I might have missed but have instructions for use been posted somewhere?  The documentation describes the visual features but how do we actually try it?

BTW the doc pages have been added to the igv-docs-website, any further changes to docs should be done there.   Currently the page is at https://github.com/igvteam/igv-docs-website/tree/main/igv-docs/docs/UserGuide/tracks/alignments/ultima.

I'm concerned about this new (to IGV) concept of "extensions".   Its not part of our current design, obviously, and I am not sure we want to take on a new concept such as this for the sake of a few visiual enhancements for data from a specific vendor.   It might be a good concept, that we would consider as part of a major refactor,  but is it required for the enhancements we are considering here?

**Comment by dror27** [comment]
> Hi all, I might have missed but have instructions for use been posted somewhere? The documentation describes the visual features but how do we actually try it?
> 
> BTW the doc pages have been added to the igv-docs-website, any further changes to docs should be done there. Currently the page is at https://github.com/igvteam/igv-docs-website/tree/main/igv-docs/docs/UserGuide/tracks/alignments/ultima.
> 
> I'm concerned about this new (to IGV) concept of "extensions". Its not part of our current design, obviously, and I am not sure we want to take on a new concept such as this for the sake of a few visiual enhancements for data from a specific vendor. It might be a good concept, that we would consider as part of a major refactor, but is it required for the enhancements we are considering here?

@jrobinso - as to the choice of implementing this visual feature using an extension modality - this has mainly been done as a the feature was developed over the past year or two - while IGV versions were moving on. The extension model allowed for isolating the code for the new visual feature into a separate set of classes, with minimal changes in IGV itself. This allowed updating the core IGV and tacking on it the visual extension.

This might not be required when the features are accepted as being part of the core IGV. Still, I can envision that some IGV users/developers might find it easier to use this extension mechanism to create new visualizations which would start as experimental and wind up in the general version. 

The bottom line is that you - as the product's architects - need to take a decision. We will probably go with what ever you decide.

As to the doc, @ilyasoifer will respond.

**Comment by ilyasoifer** [comment]
@jrobinso Data had been placed here:  `test/data/sam/flow_renderer.sam`.  Data are around chr1:10000. I believe that our extension automatically detects the Ultima-specific tags and activates the modified visualization of the indels. 

Thanks for moving the docs to the other repo, should we removed them from this PR?
Thanks!

> Hi all, I might have missed but have instructions for use been posted somewhere? The documentation describes the visual features but how do we actually try it?
> 
> BTW the doc pages have been added to the igv-docs-website, any further changes to docs should be done there. Currently the page is at https://github.com/igvteam/igv-docs-website/tree/main/igv-docs/docs/UserGuide/tracks/alignments/ultima.
> 
> I'm concerned about this new (to IGV) concept of "extensions". Its not part of our current design, obviously, and I am not sure we want to take on a new concept such as this for the sake of a few visiual enhancements for data from a specific vendor. It might be a good concept, that we would consider as part of a major refactor, but is it required for the enhancements we are considering here?

File: src/main/java/org/broad/igv/session/SessionWriter.java
Comment (jrobinso) [inline src/main/java/org/broad/igv/session/SessionWriter.java:1]: Its not clear how this change to SessionWriter is germane to this PR.

File: src/main/java/org/broad/igv/variant/VariantMenu.java
Comment (jrobinso) [inline src/main/java/org/broad/igv/variant/VariantMenu.java:1]: Is "MATEID" a standard VCF attribute?  Widely used?  Part of VCF 4.4?   We have code elsewhere that relies on "CHR2" and "END", although this is used to create a psuedo implied variant, not to link to an actual VCF record.   

I don't object to this but would like to know what it's based on.

File: src/main/java/org/broad/igv/ext/load/LoadMultipleTracks.java
Comment (jrobinso) [inline src/main/java/org/broad/igv/ext/load/LoadMultipleTracks.java:1]: I'm unclear what problem this class solves, loading multiple files (whitespace delimited)  by URL is already supported.

File: src/main/java/org/broad/igv/sam/AlignmentRenderer.java [hunk 1]
```diff
@@ -1057,6 +1082,180 @@ private void drawLargeIndelLabel(Graphics2D g, boolean isInsertion, String label
         }
     }
 
+    private void drawInsertions(Rectangle rect, Alignment alignment, RenderContext context, AlignmentTrack.RenderOptions renderOptions,
```
Comment (jrobinso) [inline src/main/java/org/broad/igv/sam/AlignmentRenderer.java]: This is a new method, most of it moved from somewhere else I think.  The github diff tool here makes it hard to see what has actually changed, and what the consequence of this new method will be, but is it neccessary in the first place?  Fewer changes are better.

**Comment by jrobinso** [comment]
@dror27 An extension framework might be a good idea in general, but we have then escalated this PR significantly.   For one thing we would need documentation on the framework itself.     Its an interesting idea but given constraints here it would be some time before we are able to consider it.

If I understand this PR the changes, from a user perspective,  amount to some additional visual elements for indels and some additions to the popup text.  I would not expect this to require many changes to code,  I think the bulk of changes here are for the extension framework.    If you could remove that (the extension) it would certainly expedite review.    An "ext" package might be renamed to "ultima" for specialized code that remains, similar to the "smrt" (pacbio) package. 

@ilyasoifer The doc can be removed from the PR, no requirement to do so,  but if you wish to make changes let's do it with PRs to the doc website.

**Comment by dror27** [comment]
@jrobinso ok. I will update.

**Comment by dror27** [comment]
@jrobinso @ilyasoifer - I have now removed the framework (concept and code). Additionally, the package has been renamed to .ultima. Thank you for your input and review

**Comment by jrobinso** [comment]
@dror27 Thanks for the quick turnaround, much easier to review now.  I had left some comments on particular files prior to this,  hopefully you can see them.  I have not used the github review function before.  2 in particular, could you explain the changes to SessionWriter and LoadFromURLAction?   I don't understand the SessionWriter change relevance, and am not sure about the "multiple tracks" addition either.  LoadFromURLAction already supports loading multiple urls, what does this change allow?

**Comment by dror27** [comment]
@jrobinso I was not able to see your individual comments, but I understand your concerns. 

The two changes mentioned (loading mutliple tracks and session-writer changes) were remains of behaviors of previous IGV versions that we patched over. They are no longer required and were therefore removed.

File: src/main/java/org/broad/igv/variant/VariantMenu.java
Comment (dror27) [reply src/main/java/org/broad/igv/variant/VariantMenu.java:1]: I have now removed this feature (Goto Mate ...) from the pull request as it is not related to the annotation and coloring of flow qualities. We will probably submit this in a separate pull request.

To answer your question, yes, it is driven by MATEID. I have no data as to how widely it is used.

File: src/main/java/org/broad/igv/sam/AlignmentRenderer.java (hunk 1 above)
Comment (dror27) [reply src/main/java/org/broad/igv/sam/AlignmentRenderer.java]: @jrobinso this method, drawInsertions, if I understand correctly, was authored outside the scope of this pull request (see Git Blame information). 

I did notice that it, along with the following methods in this class (VariantRenderer) is not used. You may consider removing them:
- drawInsertions
- renderExpandedInsertion
- getOutlierStatus
- compareToBounds

Running "unused declarations" code analysis produces many more unused methods - so I think that this is a bigger issue and might not be related to this pull request.

Comment (jrobinso) [reply src/main/java/org/broad/igv/sam/AlignmentRenderer.java]: Thanks, could you remove this method?    That will make for a cleaner merge.  Also could you remove the doc as that has been moved to the igv doc repository?  With this I think this will be ready to merge.   Thanks for your prompt responses.

**Comment by dror27** [comment]
Done - method drawInsertions and superfluous doc removed.

**Comment by jrobinso** [comment]
OK great!   I will do a pull and some quick tests and then merge, probably this evening.   Thanks for you quick responses, it makes this easier.

**Comment by jrobinso** [comment]
Thanks again for your contributions.  This is merged and will be part of the upcoming 3.0 release.    We hope to publish a preview before the end of the year, with a final release early in 2024.  If you would like to suggest a short description (1-2 sentences) for the release notes please feel free to do so here.  Otherwise we will derive a description from the documentation.

**Comment by dror27** [comment]
@jrobinso thank you for the joint work and for the merge.

@ilyasoifer can you please suggest a short description (1-2 sentences) for the release notes? (we've removed the doc from the source tree).

My take on summarizing the deleted doc:

- New graphical annotations for INDEL quality on SAM/BAM/CRAM files containing Flow Based reads. For information on Flow Based sequencing, please see https://www.biorxiv.org/content/10.1101/2022.05.29.493900

**Comment by ilyasoifer** [comment]
@jrobinso - thanks for the efficient merge. I agree with @dror27  description for the release note, just one minor tweak: 
* New graphical annotations for INDEL quality on SAM/BAM/CRAM files containing Ultima Genomics flow-based reads. For more information please see https://www.biorxiv.org/content/10.1101/2022.05.29.493900

**Comment by jrobinso** [comment]
@dror27 A problem has arose.  We need a better test for "isFlow".  The problem is "ti" and "tp" are not reserved tags, anyone can use these for any application and we now have non-Ultima data that uses "tp".   So we need something more precise.   Is there anything in the BAM header that would tell us this is a "flow" tagged file?  Any other suggestions?

Problem is reported here (or rather symptom):  https://github.com/igvteam/igv/issues/1468

```
  private boolean isFlow(SAMRecord record) {

        // must have one of the key attributes
        for ( String name : KEY_ATTR.split(",") ) {
            if ( record.hasAttribute(name) ) {
                return true;
            }
        }
        return false;
    }

**Comment by ilyasoifer** [comment]
Thanks, we will take care of this, hopefully today.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jrobinso** [comment]
@ilyasoifer Thanks, I have some ideas too but I'll wait for your suggestions.   If you branch its probably better to do so from branch "2.17", as opposed to main, but either will do.  It will be merged to both.

**Comment by dror27** [comment]
In IGV code, we also test for the "ULTIMA" value in the "PL" (platform) attribute of the associated read group, the presence for the "MC" (maxHmer) in the read group and finally the presence "tp" attribute on the record itself.

Does it make sense here?

(this assumes that we have easy access to the read group)

       final SAMReadGroupRecord readGroup = samAlignment.getRecord().getReadGroup();
        if ( !RG_ATTR_PL_ULTIMA.equals(readGroup.getAttribute(RG_ATTR_PL))
             &&  (readGroup.getAttribute(RG_ATTR_MC) == null) )
            return false;
        if ( !samAlignment.getRecord().hasAttribute(ATTR_TP)  )
            return false;

**Comment by dror27** [comment]
@jrobinso and @ilyasoifer - please see new pull request to address:
https://github.com/igvteam/igv/pull/1471

**Comment by jrobinso** [comment]
That looks pretty robust.  I'm also going to add a try/catch for any future conflicts, but with this test I doubt they'll be any.
